feat: improve task naming

jd20 opened the pull request:
A couple improvements to task naming:
* For function commands, search for first occurrence of a filename argument, and strip from that point onwards.
* For commands in general, strip common package runner prefixes (and their options): `npm`, `npx`, `yarn` (unless it would strip the entire command).
* Drop the ellipsis from function commands, to be consistent with non-function commands.

Fixes: #828

iiroj commented on lib/makeCmdTasks.js:
```diff
@@ -27,25 +27,42 @@ module.exports = async function makeCmdTasks({ commands, files, gitDir, shell })
 
-        const [startOfFn] = command.split(' ')
-        title += ` ${startOfFn} ...` // Append function name, like `[Function] eslint ...`
+      // For the title, strip package runner commands (npm/npx/yarn) from the beginning
+      if (words.length > 0 && ['npm', 'npx', 'yarn'].includes(words[0])) {
+        let idx = 1
+        while (idx < words.length && (words[idx] === '' || words[idx].startsWith('-'))) {
+          idx += 1 // Remove any options to the package runner command
+        }
+        if (idx < words.length) {
+          words.splice(0, idx) // Make sure we don't strip the entire command
+        }
```
Not sure about this. How will it format something like `(files) => npm run test -- --update ${files.join(' ')}`

jd20 replied:
It'll format it as `run test -- --update`. Do you think it makes sense to special-case the `run` command? Or maybe in general it is going to be too difficult to properly handle package runner commands, without overcomplicating things?

iiroj replied:
I think it’s best to leave out stripping those. I’d imagine there’s always some reason to run via npm/yarn instead of directly, so better not hide it.

jd20 replied:
Sounds good, I'll update the PR.

iiroj commented:
Looks nice, and a good improvement!

Let's just leave out the second bullet of stripping out prefixes to simplify code and not cause unexpected issues (users not recognizing the command).

iiroj commented:
It seems the path match doesn't quite work on Windows:

```bash
- LOG [Function] prettier --write [started]
- LOG [Function] prettier --write [completed]
+ LOG [Function] prettier --write C:\projects\lint-staged-test\AqocQtX4BMGjgyZNYR1jx\test.js [started]
+ LOG [Function] prettier --write C:\projects\lint-staged-test\AqocQtX4BMGjgyZNYR1jx\test.js [completed]
```

[Appveyor logs](https://ci.appveyor.com/project/okonet/lint-staged/builds/31970406)

iiroj commented on lib/makeCmdTasks.js:
```diff
@@ -27,25 +27,42 @@ module.exports = async function makeCmdTasks({ commands, files, gitDir, shell })
+      const words = command.trim().split(' ') // Tokenize the command into words
+
+      if (isFn) {
+        const indexOfFile = words.findIndex(word => files.includes(word)) // index of first file in the command
```
I fear this might not work as-is, since the task function can format files.

It might be necessary to do something like:

```js
const fileBasenames = files.map(file => path.basename(file))
const indexOfFile = words.findIndex(word => fileBasenames.some(basename => word.includes(basename)))
```

to match only the basename (ie. `file.js`).

This is bound to have performance indications, however.

jd20 replied:
Yeah, I also realized tokenizing by splitting on space isn't going to work properly either, if there are spaces inside the filename. Let me revise it a bit more, to try and handle the basename logic above. Is the function formatting files the reason Windows test is failing as well?

iiroj replied:
Can’t say for sure, though I would think the `files` available in this method are normalized the same way; apparently not.

iiroj commented:
Maybe the simplest method would be to, for the string `npm run test -- -u src/staged.js`, to do:

1. Get the os path separator from `path.sep`. Eg. `/`
1. Split the command string from this and get the first part, eg `npm run test -- -u src`
1. Split the above from the last space and get the first part, eg `npm run test -- -u`

It should also be try/catched to a short fallback if no path.sep is not found, so that the entire string would not end up in the console.

What do you think?

jd20 commented:
> Maybe the simplest method would be to, for the string `npm run test -- -u src/staged.js`, to do:
> 
> 1. Get the os path separator from `path.sep`. Eg. `/`
> 2. Split the command string from this and get the first part, eg `npm run test -- -u src`
> 3. Split the above from the last space and get the first part, eg `npm run test -- -u`
> 
> It should also be try/catched to a short fallback if no path.sep is not found, so that the entire string would not end up in the console.
> 
> What do you think?

Wouldn't that fail for commands with no path separators (staged files in same directory), as well as potentially strip out starting from a non-staged file, such as config?

I'm thinking to tokenize the command (taking quotes into account), then walk through each token, checking for the first one that matches to the staged set (can resolve and compare absolute paths, if we want to be certain it's same file). Then strip from that point onwards. I can optimize the check against the staged set, if we're concerned about perf.

iiroj commented:
@jd20 You are true about the failing!

fantapop commented:
@iiroj  I'm interested in this as well.  I'm Wondering if something simpler like being about to put in a manual override would be better.  I'm also wondering why the command is getting cut off in so few characters?  It seems like it could afford to go up to the width of the terminal.

iiroj commented:
@fantapop thanks for the terminal width suggestion! That seems like a pretty easy improvement.

iiroj commented:
I opened PR https://github.com/okonet/lint-staged/pull/865 which will always take the full length task title (for both "simple" and functional tasks), and then truncates it to the current terminal columns, minus the prefix padding length caused by listr2. This means the task title will be as long as possible to fit on a single line, and will be clipped with an ellipsis `…`.

I'll close this PR since that should be a good enough solution.

fantapop commented:
This solution is working great. Thank you!
